rp2040-hal/src/gpio/mod.rs

```diff
@@ -128,3 +152,1262 @@ pub enum OutputOverride {
+                    >,
+                )+
+            }
+
+            impl Pins {
+                /// Take ownership of the PAC [`PORT`] and split it into
+                /// discrete [`Pin`]s.
+                ///
+                /// This struct serves as a replacement for the HAL [`Pins`]
+                /// struct. It is intended to provide more meaningful names for
+                /// each [`Pin`] in a BSP. Any [`Pin`] not defined by the BSP is
+                /// dropped.
+                ///
+                /// [`Pin`](rp2040_hal::gpio::Pin)
+                /// [`Pins`](rp2040_hal::gpio::Pins)
+                #[inline]
+                pub fn new(io : $crate::pac::IO_BANK0, pads: $crate::pac::PADS_BANK0, sio: $crate::sio::SioGpioBank0, reset : &mut $crate::pac::RESETS) -> Self {
+                    let mut pins = $crate::gpio::Pins::new(io,pads,sio,reset);
+                    Self {
+                        $(
+                            $( #[$id_cfg] )*
+                            $name: pins.[<$Id:lower>],
+                        )+
+                    }
+                }
+            }
+            $(
+                $( #[$id_cfg] )*
+                $crate::bsp_pins!(@aliases, $( $( $( #[$alias_cfg] )* $Id $Function $PullType $Alias )+ )? );
+            )+
+        }
+    };
+    ( @aliases, $( $( $( #[$attr:meta] )* $Id:ident $Function:ident $PullType:ident $Alias:ident )+ )? ) => {
+        $crate::paste::paste! {
+            $(
+                $(
+                    $( #[$attr] )*
+                    /// Alias for a configured [`Pin`](rp2040_hal::gpio::Pin)
+                    pub type $Alias = $crate::gpio::Pin<
+                        $crate::gpio::bank0::$Id,
+                        $crate::gpio::$Function,
+                        $crate::gpio::$PullType
+                    >;
+                )+
+            )?
+        }
+    };
+}
+
+//==============================================================================
+//  InOutPin
+//==============================================================================
+
+/// A wrapper [`AnyPin`]`<Function = `[`FunctionSioOutput`]`>` emulating open-drain function.
+///
+/// This wrapper implements both InputPin and OutputPin, to simulate an open-drain pin as needed for
+/// example by the wire protocol the DHT11 sensor speaks.
+///
+/// <https://how2electronics.com/interfacing-dht11-temperature-humidity-sensor-with-raspberry-pi-pico/>
+pub struct InOutPin<T: AnyPin> {
+    inner: Pin<T::Id, FunctionSioOutput, T::Pull>,
+}
+
+impl<T: AnyPin> InOutPin<T> {
+    /// Create a new wrapper
+    pub fn new(inner: T) -> InOutPin<T>
+    where
+        T::Id: ValidFunction<FunctionSioOutput>,
+    {
+        let mut inner = inner.into();
+        inner.set_output_enable_override(OutputEnableOverride::Disable);
+
+        // into Pin<_, FunctionSioOutput, _>
+        let inner = inner.into_push_pull_output_in_state(PinState::Low);
```

Comment:
I wonder if this is the only sensible way an InOutPin could work, and therefore, if the naming is appropriate:

In theory, it would be equally valid to have pin that alternates between active drive to high and disabled driver. That could be used together with a pull-down resistor. But I have no idea if there is any practical application of this configuration, or if everything uses pull-up resistors and drive the pin low actively.

Reply:
I thought about the exact same thing as I was picking this from the dht example but I couldn't think of any usecase so I figured simpler is better and we can still eventually update this later on.

Reply:
I did some web searching about this question. It seems like having pull-up resistors and an output driver pulling the line down actually had significant advantages over the opposite choice back in the days of TTL logic. While this wouldn't matter much now that everybody is using CMOS technology, the convention has been kept (why change habits for no good reason?) and everyone uses it.

So I think it's fine to only provide this variant for now. I'd prefer a slightly more descriptive naming, but don't have a good idea. `OpenDrain` wouldn't be exactly true. Perhaps `OpenDrainLike`? But that's ugly. If we don't find anything better, I'm fine with keeping the `InOutPin` name.